power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/y_bus.hpp

```diff
@@ -330,24 +336,109 @@ template <bool sym> class YBus {
+            //  loop over all entries of this position
+            for (Idx element = y_bus_entry_indptr[entry]; element != y_bus_entry_indptr[entry + 1]; ++element) {
+                auto param_idx = y_bus_element[element].idx;
+                if (y_bus_element[element].element_type == YBusElementType::shunt) {
+                    // shunt
+                    entry_admittance += shunt_param[param_idx];
+                    entry_param_shunt.push_back(param_idx);
+                } else {
+                    // branch
+                    entry_admittance +=
+                        branch_param[param_idx].value[static_cast<Idx>(y_bus_element[element].element_type)];
+                    entry_param_branch.push_back(param_idx);
+                    // entry_param_shunt_pair.push_back(std::make_pair(param_idx, element));
+                }
+            }
+            // assign
+            admittance[entry] = entry_admittance;
+            map_admittance_param_branch.push_back(entry_param_branch);
+            map_admittance_param_shunt.push_back(entry_param_shunt);
+            // map_admittance_param_shunt_pairs.push_back(entry_param_shunt_pair);
+        }
+        // move to shared ownership
+        admittance_ = std::make_shared<ComplexTensorVector<sym> const>(std::move(admittance));
+    }
+
+    IdxVector increments_to_entries() {
+        // construct affected entries
+        std::set<Idx> affected_entries;
+        auto querry_params_in_map = [&affected_entries, this](auto params_to_change, bool is_branch) {
+            if (is_branch) {
+                for (size_t i = 0; i < map_admittance_param_branch.size(); ++i) {
+                    if (std::any_of(map_admittance_param_branch[i].begin(), map_admittance_param_branch[i].end(),
+                                    [&](Idx val) {
+                                        return std::find(params_to_change.begin(), params_to_change.end(), val) !=
+                                               params_to_change.end();
+                                    })) {
+                        affected_entries.insert(static_cast<int>(i));
+                    }
+                }
+            } else {
+                for (size_t i = 0; i < map_admittance_param_shunt.size(); ++i) {
+                    if (std::any_of(map_admittance_param_shunt[i].begin(), map_admittance_param_shunt[i].end(),
+                                    [&](Idx val) {
+                                        return std::find(params_to_change.begin(), params_to_change.end(), val) !=
+                                               params_to_change.end();
+                                    })) {
+                        affected_entries.insert(static_cast<int>(i));
+                    }
+                }
+            }
+        };
+
+        querry_params_in_map(math_model_param_incrmt_->branch_param_to_change, true);
+        querry_params_in_map(math_model_param_incrmt_->shunt_param_to_change, false);
```

Comment:
combining logic is only useful when there's no additional differences. I would instead extract the common part, e.g.

```suggestion
        auto querry_params_in_map = [&affected_entries, this](auto const& params_to_change, auto const& mapping) {
            for (size_t i = 0; i < mapping.size(); ++i) {
                if (std::ranges::any_of(mapping[i],
                                        [&](Idx val) { return std::ranges::contains(params_to_change, val); })) {
                    affected_entries.insert(static_cast<int>(i));
                }
            }
        };

        querry_params_in_map(math_model_param_incrmt_->branch_param_to_change, map_admittance_param_branch);
        querry_params_in_map(math_model_param_incrmt_->shunt_param_to_change, map_admittance_param_shunt);
```

Reply:
Well, now that I came back to one of my staged code I realize why I put them separate in the first place; for the branch param I wanted to keep track of not only the index but also the entry in param[index] that changed. It turn out to be not what I expected, still wip.

Reply:
Also noticed that perhaps `std::ranges::contains` is not properly supported yet, either. Stick to `find` for now.